src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs

```diff
@@ -7752,6 +7752,295 @@ void M()
+            comp.VerifyDiagnostics();
+            var verifier = CompileAndVerify(comp, expectedOutput: "1 2");
+            verifier.VerifyIL("C.M", @"{
+  // Code size       27 (0x1b)
+  .maxstack  3
+  .locals init (S V_0,
+                int V_1,
+                int V_2)
+  IL_0000:  nop
+  IL_0001:  ldarg.0
+  IL_0002:  ldfld      ""S C.field""
+  IL_0007:  stloc.0
+  IL_0008:  ldloca.s   V_0
+  IL_000a:  ldloca.s   V_1
+  IL_000c:  ldloca.s   V_2
+  IL_000e:  call       ""void Extension.Deconstruct(in S, out int, out int)""
+  IL_0013:  nop
+  IL_0014:  ldarg.1
+  IL_0015:  ldloc.1
+  IL_0016:  stind.i4
+  IL_0017:  ldarg.2
+  IL_0018:  ldloc.2
+  IL_0019:  stind.i4
+  IL_001a:  ret
+}");
+        }
+
+        [Fact, WorkItem(24160, "https://github.com/dotnet/roslyn/issues/24160")]
+        public void TestInThisDeconstruct_Nested()
+        {
+            var source = @"
+public class C
+{
+    static S field = default;
+    static void Main()
+    {
+        var (x, (y, z)) = (1, field);
+        System.Console.Write($""{y} {z}"");
+    }
+}
+public struct S { }
+public static class Extension
+{
+    public static void Deconstruct(in this S d, out int x1, out int x2)
+    {
+        x1 = 1;
+        x2 = 2;
+    }
+}
+";
+
+            var comp = CreateCompilationWithMscorlib45(source, options: TestOptions.DebugExe, references: s_valueTupleRefs);
+            comp.VerifyDiagnostics();
+            CompileAndVerify(comp, expectedOutput: "1 2");
+        }
+
+        [Fact, WorkItem(24160, "https://github.com/dotnet/roslyn/issues/24160")]
+        public void TestInThisDeconstruct_Foreach()
+        {
+            var source = @"
+public class C
+{
+    static S field = default;
+    static void Main()
+    {
+        foreach (var (y, z) in new[] { field })
+        {
+            System.Console.Write($""{y} {z}"");
+        }
+    }
+}
+public struct S { }
+public static class Extension
+{
+    public static void Deconstruct(in this S d, out int x1, out int x2)
+    {
+        x1 = 1;
+        x2 = 2;
+    }
+}
+";
+
+            var comp = CreateCompilationWithMscorlib45(source, options: TestOptions.DebugExe, references: s_valueTupleRefs);
+            comp.VerifyDiagnostics();
+            CompileAndVerify(comp, expectedOutput: "1 2");
+        }
+
+        [Fact, WorkItem(24160, "https://github.com/dotnet/roslyn/issues/24160")]
+        public void TestRefThisDeconstruct()
+        {
+            var source = @"
+public class C
+{
+    S field = default;
+    static void Main()
+    {
+        new C().M();
+    }
+    void M()
+    {
+        var (x, y) = field;
+        System.Console.Write($""{field.touched} {x + y}"");
+    }
+}
+public struct S
+{
+    public bool touched;
+}
+public static class Extension
+{
+    public static void Deconstruct(ref this S d, out int x1, out int x2)
+    {
+        // d is a reference to a temp used for deconstruction
```

Comment:
>// d is a reference to a temp used for deconstruction [](start = 8, length = 53)

Is this consistent with behavior when Deconstruct is an instance method?

Reply:
Yes. In all invocations of `Deconstruct` produced during lowering, the target of the invocation is a saved value in a temp, if it was side-effecting (`EvaluateSideEffectingArgumentToTemp`).